mono/mini/mini-windows.c

```diff
@@ -109,57 +111,75 @@ win32_time_proc (UINT uID, UINT uMsg, DWORD dwUser, DWORD dw1, DWORD dw2)
 	}
 }
 
-void
-mono_runtime_setup_stat_profiler (void)
+static VOID CALLBACK
+timer_event_proc (UINT uID, UINT uMsg, DWORD_PTR dwUser, DWORD_PTR dw1, DWORD_PTR dw2)
 {
-	static int inited = 0;
-	TIMECAPS timecaps;
+	thread_timer_expired ((HANDLE)dwUser);
+}
 
-	if (inited)
-		return;
+static VOID
+stop_profiler_timer_event (void)
+{
+	if (g_timer_event != 0) {
+
+		timeKillEvent (g_timer_event);
+		g_timer_event = 0;
+	}
+
+	if (g_timer_main_thread != INVALID_HANDLE_VALUE) {
+
+		CloseHandle (g_timer_main_thread);
+		g_timer_main_thread = INVALID_HANDLE_VALUE;
+	}
+}
+
+static VOID
+start_profiler_timer_event (void)
+{
+	g_return_if_fail (g_timer_main_thread == INVALID_HANDLE_VALUE && g_timer_event == 0);
+
+	TIMECAPS timecaps;
 
-	inited = 1;
 	if (timeGetDevCaps (&timecaps, sizeof (timecaps)) != TIMERR_NOERROR)
 		return;
 
-	if ((win32_main_thread = OpenThread (READ_CONTROL | THREAD_GET_CONTEXT, FALSE, GetCurrentThreadId ())) == NULL)
+	g_timer_main_thread = OpenThread (READ_CONTROL | THREAD_GET_CONTEXT, FALSE, GetCurrentThreadId ());
+	if (g_timer_main_thread == NULL)
 		return;
 
 	if (timeBeginPeriod (1) != TIMERR_NOERROR)
 		return;
 
-	if ((win32_timer = timeSetEvent (1, 0, (LPTIMECALLBACK)win32_time_proc, (DWORD_PTR)NULL, TIME_PERIODIC)) == 0) {
+	g_timer_event = timeSetEvent (1, 0, (LPTIMECALLBACK)timer_event_proc, (DWORD_PTR)g_timer_main_thread, TIME_PERIODIC | TIME_KILL_SYNCHRONOUS);
+	if (g_timer_event == 0) {
 		timeEndPeriod (1);
 		return;
 	}
 }
 
 void
+mono_runtime_setup_stat_profiler (void)
+{
+	start_profiler_timer_event ();
```

Comment:
@alexrp Looks like we're getting some patches towards making the profiler works on windows :)